### PR TITLE
fix: normalize original sourcemap sources

### PR DESCRIPTION
**Pull request by underfin**

### CODE REVIEW

File: packages/rolldown/src/plugin/bindingify-build-hooks.ts [hunk 1]
```diff
@@ -119,6 +124,24 @@ export function bindingifyLoad(
...
+      return { code: ret.code }
+    }
+
+    let map =
+      typeof ret.map === 'object'
+        ? ret.map
+        : (JSON.parse(ret.map) as SourceMapInputObject)
+    if (!isEmptySourcemapFiled(map.sources)) {
+      // normalize original sourcemap sources
+      // Port form https://github.com/rollup/rollup/blob/master/src/utils/collapseSourcemaps.ts#L180-L188.
```
Comment (hyf0) [inline packages/rolldown/src/plugin/bindingify-build-hooks.ts:137]: Is this a temporary workaround? Is the practical way of doing this in rust's `collapseSourcemaps` function?

Comment (underfin) [reply packages/rolldown/src/plugin/bindingify-build-hooks.ts:137]: The rust `collapseSourcemaps` is a stable sourcemap remapping.  Here normalize sources path to support rollup cases, it is unrelevant to remapping.

Comment (hyf0) [reply packages/rolldown/src/plugin/bindingify-build-hooks.ts:137]: I saw this is for test https://github.com/rollup/rollup/blob/64870c31062983a741832d8cd2d4bcdf55e4de55/test/sourcemaps/samples/loaders/_config.js.

What happen if we have the same logic in rust plugins? 

I saw the test is to ensure the absolute path become relative. So I guess this is the normalization this PR does. Is my understanding right?

image

image

**Comment by hyf0** [comment]
I still have questions, but let's merge it first.

---

https://github.com/rolldown/rolldown/pull/765#discussion_r1554556939

**Review by hyf0 — APPROVED** [review]

File: packages/rolldown/src/plugin/bindingify-build-hooks.ts (hunk 1 above)
Comment (underfin) [reply packages/rolldown/src/plugin/bindingify-build-hooks.ts:137]: Here it is intend to test `sourceRoot` feature. You should see here https://github.com/rollup/rollup/blob/64870c31062983a741832d8cd2d4bcdf55e4de55/test/sourcemaps/samples/loaders/_config.js#L27-L30.

Comment (underfin) [reply packages/rolldown/src/plugin/bindingify-build-hooks.ts:137]: > What happen if we have the same logic in rust plugins?

The logic maybe should be put into rust side, but I'm not sure at now. It need to loader `sourcemap` is stable, the plugin should normalize `sourceRoot`.

**Pull request merged**